### PR TITLE
[Just In Time Messages] Add build and locale identifiers to JITM request

### LINKED ISSUE #8015: [Just In Time Message] Add build identifiers to query string for JITMs
## Description
In Project Salas, we've added the ability to send Just In Time Messages to all Woo Mobile users, initially displayed on the dashboard. A single message definition will be picked up both on iOS and Android. This is beneficial as we don't need to define the same messages multiple times for different platforms...

However, there will definitely be cases when we want to send a message to a smaller subset of our users, or define test JITMs which are not yet ready for production. The `with_query_string` and `without_query_string` JITM rules can help here, but only if we pass useful data in the query.

These rules were intended to work on the query string in a browser's url bar, when accessed from PHP web code. Since we request JITMs through the REST API, we also need to pass the "query string" in the request. The JITM engine doesn't check the REST request's query string for these parameters.

See pdfdoF-1Fe-p2 for full context.

## Locale

`locale=en_US` or other phone locale code. This is unused and for future proofing. It should be added at the same level as the named `path` and `query` URL Query Parameters (see example request below.)

## Query parameters to add:

- `platform` = `ios`
- `version` = `11.1`: use `Bundle.main.marketingVersion`. This changes to e.g. `11.1.1` for hot fix builds.
- `build_type` = `developer` for `localDeveloper` and CI builds. Use `BuildConfiguration.current.rawValue`, in combination with the [technique in AppRatingsManager](https://github.com/woocommerce/woocommerce-ios/blob/b5234ea1fc1312ddaf99056f285d6afc6d444f38/WooCommerce/Classes/Tools/AppRatings/AppRatingManager.swift#L31) to determine TestFlight beta builds. N.B. don't include this parameter on alpha/beta/production builds.
- `device` = `phone` or `pad`. Use `UIDevice.current.userInterfaceIdiom.rawValue == 0` for `phone`, and `1` for `pad`. Don't include this parameter for other values.

## Request format:

These JITM parameters should be constructed as a query string, before inclusion in the JSON payload of the `GET /jetpack/v4/jitm/` request. This is unlike the locale parameter, which is a top level URL query parameter. Be careful to check all this, because the name `query` is used in several places!

### Example query string:

- Query string: `platform=ios&version=11.1&build_type=developer&device=phone`
- JSON: `{"message_path": "woomobile:my_store:admin_notices", "query": "platform=ios&version=11.1&build_type=developer&device=phone"}`
- URL Query Param: `query=%7B%22message_path%22%3A%20%22woomobile%3Amy_store%3Aadmin_notices%22,%20%22query%22%3A%20%22platform%3Dios%26version%3D11.1%26build_type%3Ddeveloper%26device%3Dphone%22%7D`

This will lead to a request like this:

```
curl "https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/{SITE_ID}/rest-api?path=jetpack%2Fv4%2Fjitm&query=%7B%22message_path%22%3A%20%22woomobile%3Amy_store%3Aadmin_notices%22,%20%22query%22%3A%20%22platform%3Dios%26version%3D11.1%26build_type%3Ddeveloper%26device%3Dphone%22%7D&locale=en_US" \
     -H 'Host: public-api.wordpress.com' \
     -H 'Authorization: Bearer {TOKEN}' \
     -H 'Content-Type: application/x-www-form-urlencoded; charset=utf-8'
```

### PR DESCRIPTION
**Pull request by joshheald**
Closes: #8015

## Description

In Project Salas, we've added the ability to send Just In Time Messages to all Woo Mobile users, initially displayed on the dashboard. A single message definition will be picked up both on iOS and Android. There are also cases when we want to send a message to a smaller subset of our users, or define test JITMs which are not yet ready for production. The `with_query_string` and `without_query_string` JITM rules can help here, but only if we pass useful data in the query.

### JITM Query
To take advantage of these rules, this PR adds the following to the GET `/jetpack/v4/jitm` request, in the `query` part of the `query` object, which is passed to the JITM engine:

- `platform` = `ios`
- `version` = `11.1`
- `build_type` = `developer` for `localDeveloper` and CI builds. N.B. we don't include this parameter on alpha/beta/production builds.
- `device` = `phone` or `pad`

### Locale
There is undocumented support for a `locale` query parameter (top level) on our API. This is not supported in the JITM endpoints at present, but passing it now will allow us to support locale mismatch between the site and phone in future. At present, all JITMs will be localised to the site or user's default language.

`locale=en_US` or other phone locale code. This deliberately excludes any currency or other locale identifier information.

See pdfdoF-1Fe-p2 for full context.

## Testing instructions

Configure Proxyman/Charles to view your network traffic, or use the WormHoly tools in the settings menu to view the requests later.

1. Run the app
2. Observe that a request is made to `/jetpack/v4/jitm`
3. Observe that the request has a `locale` parameter with your phone's locale as the value, at the top level of the request – `1` in screenshot below
4. Observe that the request has a `query` parameter, with a JSON object, with an internal `query` key.
5. Observe that the value for that key has `platform`, `version`, `build_type`, and `device` values as specified above – `2` in screenshot below.

N.B. the values for these parameters, including locale, are not currently used for anything by the backend.

## Screenshots

CleanShot 2022-11-03 at 14 28 53@2x

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Experiments/Experiments/DefaultFeatureFlagService.swift [hunk 1]
```diff
@@ -48,7 +48,7 @@ public struct DefaultFeatureFlagService: FeatureFlagService {
         case .simplifiedLoginFlowI1:
             return buildConfig == .localDeveloper || buildConfig == .alpha
         case .justInTimeMessagesOnDashboard:
-            return buildConfig == .localDeveloper || buildConfig == .alpha
+            return true
```
Comment (koke) [inline Experiments/Experiments/DefaultFeatureFlagService.swift:43]: Was this meant to go in this PR?

**Review by koke — APPROVED** [review]
Works as expected and code looks good. However the last few commits seem more related to the JITM public release than the description of this PR. Are these actually intended for this branch?

File: Experiments/Experiments/DefaultFeatureFlagService.swift (hunk 1 above)
Comment (joshheald) [reply Experiments/Experiments/DefaultFeatureFlagService.swift:43]: Yep, that's intentional, this is the last bit required for the release of i1. I've checked that it works correctly on a release build.

Should have called it out in the description, sorry!

**Comment by joshheald** [comment]
> Works as expected and code looks good.

@koke thanks again for the review!

**Review by pachlava — APPROVED** [review]
@joshheald 👋 

I've checked that build and locale identifiers now show up for iPhone:

Screenshot 2022-11-04 at 13 11 06

And that they change for iPad / other locale:

SCR-20221104-inh

One small question: is it correct that when the app is not localized for the device locale in use (e.g. Ukrainian), the value still stays `en_US`?

**Comment by joshheald** [comment]
@pachlava thanks as always for the review and detailed testing!

> One small question: is it correct that when the app is not localised for the device locale in use (e.g. Ukrainian), the value still stays `en_US`?

I'd not tested this case, but if that's the behaviour it makes sense for our use case. If the reverse were to happen, we would have a localised message in an otherwise unlocalised app, which would seem pretty strange.

I didn't code that deliberately: I use `Locale.current` from the system, which must return the app's locale as `en_US` as its fallback. Because of this, it should also support hierarchies of localisation preferences.

**Comment by joshheald** [comment]
During testing with the new JITM testing definitions, I noticed that the display of JITMs had become inconsistent. Sometimes they showed up, sometimes they didn't.

This was because the `build_type` parameter is ignored when it's not the first parameter in the query string. That _appears_ to be a backend bug, and I'll follow it up to get it logged and fixed.

In the meantime, I've added a workaround in 0ac84df, by sorting the query params with `build_type` and `platform` first: the two params we are most likely to care about initially.

**Pull request merged**